File: 2024/albert-arvidsson-polyglot/day_05/part_1/solve.rb
#!/usr/bin/env ruby

rules, updates = File.read(ARGV[0]).split("\n\n").map { |s| s.split("\n") }
rules = rules.to_set

manuals = updates.map { |s| s.split(',') }.filter do |update|
  update == update.sort do |a, b|
    next -1 if rules.include?("#{a}|#{b}")

    rules.include?("#{b}|#{a}") ? 1 : 0
  end
end

pp(manuals.sum { |manual| manual[manual.size / 2].to_i })

File: 2024/albert-arvidsson-polyglot/day_05/part_1/solve.mjs
#!/usr/bin/env node

import { readFileSync } from "fs";

const input = readFileSync(process.argv[2], "utf-8").split("\n\n");
const set = new Set(input[0].split("\n"));

const manuals = input[1]
  .trim()
  .split("\n")
  .map((s) => s.split(","))
  .filter((update) => {
    const sorted = update
      .slice()
      .sort((a, b) =>
        set.has(`${a}|${b}`) ? -1 : set.has(`${b}|${a}`) ? 1 : 0,
      );
    return sorted.join(",") === update.join(",");
  });

console.log(manuals.reduce((acc, a) => acc + +a[Math.floor(a.length / 2)], 0));
